src/main/java/com/scalar/cassy/transferer/AzureBlobFileDownloader.java

```diff
@@ -1,63 +1,88 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import reactor.core.publisher.Mono;
 
 public class AzureBlobFileDownloader implements FileDownloader {
   private static final Logger logger = LoggerFactory.getLogger(AzureBlobFileDownloader.class);
-  private BlobContainerAsyncClient blobContainerClient;
+  private BlobContainerClient blobContainerClient;
+  private static final int NUM_THREADS = 3;
+  private final ExecutorService executorService = Executors.newFixedThreadPool(NUM_THREADS);
 
   @Inject
-  public AzureBlobFileDownloader(BlobContainerAsyncClient blobContainerClient) {
+  public AzureBlobFileDownloader(BlobContainerClient blobContainerClient) {
     this.blobContainerClient = blobContainerClient;
   }
 
   @Override
   public void download(RestoreConfig config) {
+    List<Future> downloadFuture = new ArrayList<>();
     String key = BackupPath.create(config, config.getKeyspace());
 
     logger.info("Downloading " + blobContainerClient.getBlobContainerName() + "/" + key);
-    List<Mono<BlobProperties>> filesToBeDownloaded = new ArrayList<>();
-    Iterable<BlobItem> keyspaceBlobs =
-        blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/")).toIterable();
-    for (BlobItem blob : keyspaceBlobs) {
+    Iterator<BlobItem> keyspaceBlobs =
+            blobContainerClient.listBlobs(new ListBlobsOptions().setPrefix(key + "/"), null).iterator();
+    while (keyspaceBlobs.hasNext()) {
+      BlobItem blob = keyspaceBlobs.next();
       Path destFile = Paths.get(config.getDataDir(), blob.getName());
       try {
         Files.createDirectories(destFile.getParent());
       } catch (IOException e) {
         throw new FileTransferException(e);
       }
-      filesToBeDownloaded.add(
-          blobContainerClient
-              .getBlobAsyncClient(blob.getName())
-              .downloadToFile(destFile.toString())
-              .doOnSuccess(
-                  blobProperties -> logger.info("Download file succeeded : " + destFile.toString()))
-              .doOnError(
-                  error -> {
-                    throw new FileTransferException(
-                        "Download file failed : " + destFile.toString(), error);
-                  }));
+      downloadFuture.add(
+          executorService
+              .submit(() -> {
+                try {
+                  try (OutputStream outputStream = writeStream(destFile)) {
+                    blobContainerClient
+                        .getBlobClient(blob.getName())
+                        .download(outputStream);
+                  }
+                } catch (IOException e) {
+                  throw new FileTransferException(e);
+                }
+                logger.info("Download file succeeded : " + destFile.toString());
+              }));
     }
 
-    // Start asynchronously all downloads and wait for all of them to complete
-    Mono.when(filesToBeDownloaded).block();
+    downloadFuture.forEach(d->{
```

Comment:
It looks like it keeps adding Future to downloadFuture and downloadFuture doesn't seem to be cleared. It doesn't seem to be OK, but is it?
If it is not OK, then you never tested with more than 3 files?

Reply:
This implementation is working fine for me.
I tested it in an Azure PSIM environment with more than 65 GB of data.

Do I need to change the implementation?